design/als-design.md

```diff
@@ -0,0 +1,68 @@
+# AccessLog Service (ALS) Support
+
+Status: Draft
+
+## Abstract
+
+This document proposed a design for supporting Envoy's AccessLog Service capabilities in Contour.
+
+## Background
+
+Envoy has the ability to send rich access logs to access logging sinks.
+With AccessLog Service (ALS) enabled, Envoy can send access log messages using gRPC streams to a gRPC access log service. 
+
+This document focuses on configuring access log sinks and enabling Envoy to send access log messages via Contour.
+
+## Goals
+- Enable users to specify ALS configuration
+- Enable users to send HTTP access logs for specific listeners 
+
+## Non Goals
+- TCP Access Log Support
+
+## High-Level Design
+Contour will add support for Envoy's HTTP Access Log Service.
+ALS information will be specified as a static cluster for Envoy and Contour controller will set `access_log` field for each HTTPConnectionManager filter.
+
+## Detailed Design
+To enable Envoy to send access logs to a sink; first a static cluster needs to be set for Envoy.
+There are two ways to add this configuration:
+- ExtensionService
+- Contour Configuration file
+
+### ExtensionService
+Contour administrator/operator can add an ExtensionService with ALS configuration details.
+As part of `contour bootstrap`; the ExtensionService used to configure a static ALS cluster for Envoy.
+Contour controller will set `access_log` field for each HTTP filter.
+
+An annotation `projectcontour.io/als: true` needs to be added to the particular ExtensionService.
+This annotations hints `contour bootstrap` to use this ExtensionService for setting the ALS cluster and Contour controller to use this ALS cluster for populating `access_log` field for each HTTP filter.
+For example:
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ExtensionService
+metadata:
+  namespace: projectcontour
+  name: accesslog-service
+  annotations:
+    projectcontour.io/als: true
+spec:
+  protocol: h2
+  services:
+    - name: accesslog-service
+      port: 8080
+```
+### Configuration File
+`contour bootstrap` and `contour serve` commmands can read from a configuration file which is mounted via a Kubernetes Configmap.
```

Comment:
I don't think we currently are able to pass a config file to `contour bootstrap` so everything is currently done with CLI flags and arguments.

Reply:
fixed.